sisyphus/prowAccessor.go

```diff
@@ -0,0 +1,161 @@
+	gubernatorURL string
+	gcsClient     *u.GCSClient
+}
+
+// NewProwAccessor creates a new ProwAccessor
+func NewProwAccessor(prowProject, prowZone, gubernatorURL, gcsBucket string) *ProwAccessor {
+	return &ProwAccessor{
+		prowProject:   prowProject,
+		prowZone:      prowZone,
+		gubernatorURL: gubernatorURL,
+		gcsClient:     u.NewGCSClient(gcsBucket),
+	}
+}
+
+// GetLatestRun reads latest run from "latest-build.txt" of a job under the gcs bucket
+// Example: {gcsBucket}}/{jobName}/latest-build.txt
+func (p *ProwAccessor) GetLatestRun(jobName string) (int, error) {
+	lastBuildFile := filepath.Join(jobName, lastBuildTXT)
+	latestBuildString, err := p.gcsClient.Read(lastBuildFile)
+	if err != nil {
+		return 0, err
+	}
+	latestBuildInt, err := strconv.Atoi(latestBuildString)
+	if err != nil {
+		log.Printf("Failed to convert %s to int: %v\n", latestBuildString, err)
+		return 0, err
+	}
+	return latestBuildInt, nil
+}
+
+// GetProwResult returns the ProwResult of the job at a specific run
+func (p *ProwAccessor) GetProwResult(jobName string, runNo int) (*ProwResult, error) {
+	jobFinishedFile := filepath.Join(jobName, strconv.Itoa(runNo), finishedJSON)
+	prowResultString, err := p.gcsClient.Read(jobFinishedFile)
+	if err != nil {
+		log.Printf("Cannot access %s on GCS: %v", jobFinishedFile, err)
+		return nil, err
+	}
+	prowResult := ProwResult{}
+	if err = json.Unmarshal([]byte(prowResultString), &prowResult); err != nil {
+		log.Printf("Failed to unmarshal ProwResult %s: %v", prowResultString, err)
+		return nil, err
+	}
+	return &prowResult, nil
+}
+
+// GetGubernatorURL returns the gubernator URL used by this ProwAccessor
+func (p *ProwAccessor) GetGubernatorURL() string {
+	return p.gubernatorURL
+}
+
+// Rerun starts on Prow the reruns on specified jobs
+func (p *ProwAccessor) Rerun(jobName string, runNo, numRerun int) error {
+	cfg, err := p.getProwJobConfig(jobName, runNo)
+	if err != nil {
+		return err
+	}
+	if err = p.triggerConcurrentReruns(jobName, cfg.Node, numRerun); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getProwJobConfig fetches the config of the job at runNo
+func (p *ProwAccessor) getProwJobConfig(jobName string, runNo int) (*ProwJobConfig, error) {
+	jobStartedFile := filepath.Join(jobName, strconv.Itoa(runNo), startedJSON)
+	StartedFileString, err := p.gcsClient.Read(jobStartedFile)
+	if err != nil {
+		return nil, err
+	}
+	cfg := ProwJobConfig{}
+	if err = json.Unmarshal([]byte(StartedFileString), &cfg); err != nil {
+		log.Printf("Failed to unmarshal ProwJobConfig %s: %v\n", StartedFileString, err)
+		return nil, err
+	}
+	return &cfg, nil
+}
+
+func (p *ProwAccessor) triggerConcurrentReruns(jobName, node string, numRerun int) error {
+	log.Printf("Rerunning %s\n", jobName)
+	recess := 1 * time.Minute
+	maxRetry := 3
+	for i := 0; i < numRerun; i++ {
+		if err := u.Retry(recess, maxRetry, func() error {
+			_, e := u.Shell("kubectl create -f \"https://prow.istio.io/rerun?prowjob=%s\"", node)
```

Comment:
Same comment here, you might want to abstract this call to ease testing of the rest of the code

Reply:
Done